src/basic/time-util.c

```diff
@@ -890,24 +880,93 @@ static int parse_timestamp_impl(const char *t, usec_t *ret, bool with_tz) {
         return 0;
 }
 
+static int parse_timestamp_with_tz(const char *t, size_t len, bool utc, int isdst, long gmtoff, usec_t *ret) {
+        _cleanup_free_ char *buf = NULL;
+
+        assert(t);
+        assert(len > 0);
+
+        buf = strndup(t, len);
+        if (!buf)
+                return -ENOMEM;
+
+        return parse_timestamp_impl(buf, /* with_tz = */ true, utc, isdst, gmtoff, ret);
+}
+
+static int parse_timestamp_maybe_with_tz(const char *t, size_t len, bool valid_tz, usec_t *ret) {
```

Comment:
hmm "len" is a bit of a misnomer, it's after all *not* length of `t`, but just some offset into it where the tz likely is...

maybe rename tz_offset? or tz_index? or tz_pos?